src/wms/src/main/java/org/geoserver/wms/WMS.java

```diff
@@ -324,6 +324,77 @@ public String getVersion() {
         return version;
     }
 
+    /**
+     * Checks if request would result in a number of dimensions that exceeds the configured maximum
+     *
+     * @param mapLayerInfo the layer info
+     * @param times the times requested
+     * @param elevations the elevations requested
+     * @param isCoverage true if the layer is a coverage
+     * @throws IOException if an error occurs
+     */
+    public void checkMaxDimensions(
+            MapLayerInfo mapLayerInfo,
+            List<Object> times,
+            List<Object> elevations,
+            boolean isCoverage)
+            throws IOException {
+        if (getServiceInfo() == null
+                || getServiceInfo().getMaxRequestedDimensionValues()
+                        == DimensionInfo.DEFAULT_MAX_REQUESTED_DIMENSION_VALUES) {
+            // max is default, which is unlimited
+            return;
+        }
+        TreeSet<Object> treeSet = new TreeSet<>();
+        int maxDimensionsToTest = getServiceInfo().getMaxRequestedDimensionValues();
+        if (times != null && !times.isEmpty() && times.get(0) instanceof DateRange) {
+            // list is null for default value, contains Date(s) for individual values, we are only
+            // checking range here
+            DateRange range = (DateRange) times.get(0);
```

Comment:
It's a bit optimistic, one can provide a comma-separated list of single points and ranges (various of them). E.g., for elevation if could be something like, ``10,15/30,25,5/17,100/200`` (e.g. mix of types, multiple ranges, eventually out of order). So this needs to loop over the values, add the single values in the set, grab the range contents, and check the limit at the end of each loop for good measure.